src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -180,3 +189,65 @@ def check_write_limits(self, use_case_id: UseCaseKey, keys: KeyCollection) -> Ra
 
 
 writes_limiter = WritesLimiter()
+
+
+class WritesLimitingStringIndexerDecorator(StringIndexer):
+    def __init__(self, indexer: StringIndexer):
+        self.indexer = indexer
+
+    def bulk_resolve(
+        self, use_case_id: UseCaseKey, org_strings: Mapping[int, Set[str]]
+    ) -> KeyResults:
+        return self.indexer.bulk_resolve(use_case_id, org_strings)
+
+    def bulk_record(
+        self, use_case_id: UseCaseKey, org_strings: Mapping[int, Set[str]]
+    ) -> KeyResults:
+        if not org_strings:
+            return KeyResults()
+
+        # we do a double-take on which strings already exist, so we know which
+        # ones to rate limit. hopefully this call does not go directly to
+        # postgres but a memcached first.
+        db_read_keys = KeyCollection(org_strings)
+        db_read_key_results = self.indexer.bulk_resolve(use_case_id, org_strings)
+        db_write_keys = db_read_key_results.get_unmapped_keys(db_read_keys)
```

Comment:
I don't think it's the rate limiters job to worry about which strings can be resolved from cache, nor require that a StringIndexerCache be passed in here for reasonable performance. I think we should flip the order of the cache and the rate limiter so that by the time strings make it here all the ones we got cache hits for are excluded. Then we shouldn't need this call to `indexer.bulk_resolve()` from this `bulk_record()` method.

Reply:
Agreed. While there is a sequential coupling between resolve and record on the rate limiter: first resolve, then rate limit, then record. The cache is orthogonal to this, there is no value in having it in the middle, it makes things harder to reason with in isolation as now we need to think about the cache when trying to follow the logic.

Reply:
Taking a step back though, I am not convinced the rate limiter fits the Indexer abstraction. And this is also the reason why you had to change the interface.
The indexer implementations (postgres, static, potentially caching) do two things in bulk record:
- resolve what they can locally, 
- either write or delegate what they cannot deal with to the following step.

They all represent a form of storage.
The Rate limiter one is not an indexer except for the fact that it follows the indexer interface. It does not represent a storage.
It does not resolve anything. Instead it relies on something else to resolve string and then applies a constraint.

This is the underlying reason why you had to break down the bulk record method into bulk_resolve and bulk_record, which makes the interface more complex to reason with. Also this break down introduce a constraint for all implementations that is artificial (it is due to the rate limiter implementation and not to a design choice). The constraint is that read and write cannot be atomic.

What if instead we modeled the cache as an Indexer and keep the rate limiter as a separate abstraction via the delegator you already have ? The duplication when introducing the rate limiter would be minimal.

Reply:
If you had the Cache follow the indexer implementation you could have the Cache doing something like the following? Assuming im following the abstraction correctly


```python
# cache's bulk_record
def bulk_record()
     # ...cache logic to look up 
    db_results = self.indexer.bulk_record(db_read_keys) # ratelimiter abstraction included in the db indexer

    filtered_db_results = filter_ratelimited_results(db_results) # remove results that don't have an id

    # ...cache logic to set values in the cache using filter_db_results
    return db_results
```